【サーバーサイド】商品購入機能

nissy7ok opened the pull request:
# What
商品購入機能を実装する。
PAY.JPのAPIを利用して
- クレジットカードの登録
- クレジットカードによる支払い
を可能にする。

# Why
フリマアプリとして必須の機能であるため。

Commit:
create pull request

Commit:
_template.scssの修正

本番環境対応時の修正漏れ

Commit:
Merge remote-tracking branch 'origin' into develop_function_of_purchase

Commit:
fix templates.scss

Commit:
テーブル作成

Commit:
fixup! テーブル作成

Commit:
fixup! fixup! テーブル作成

Commit:
商品情報の表示

Commit:
fixup! 商品情報の表示

Commit:
fixup! 商品情報の表示

Commit:
URLに商品IDを含める

URLに記述されたIDによって表示内容も切り替える

Commit:
ビューの修正

- ヘッダー・フッター部のロゴをトップページへのリンクに
- フッター部にリスト追加(レスポンシブ対応)
- 「変更する」リンクの色を変更

Commit:
devise導入

Commit:
ユーザー認証実装

認証機能実装のため標準のログイン画面も実装。
ビューのマークアップや細かいサーバーサイドの実装は別で行う。

Commit:
ビューの修正

郵便番号をにハイフンを挿入

Commit:
gem 'payjp'の導入

cardモデル、コントローラーの作成

Commit:
gem 'dotenv-rails'の導入

環境変数を管理するgemであるdotenv-railsを導入。
.envファイルに秘密鍵を記入しgitignoreに指定。
理由：PAY.JP APIの秘密鍵をプッシュしないため。
秘密鍵が入ったファイルを後日渡します。

Commit:
cardモデルの修正

Commit:
カード登録フォームを仮作成

カード登録フォームを仮で作成。
トークンを取得するところまで実装済み。
ここから取得したトークンをDBに保存できるようにする。
ビューは機能完成後に整える。

Commit:
カード登録機能実装

カード情報をPAY.JPのAPIに送信して、返ってきた顧客IDとカードIDをDBに保存できるようにした。

Commit:
fixup! カード登録機能実装

Commit:
購入確認画面でのカード情報の表示機能実装

その他アクセス制限も追加
-move_to_login:ログインしていないユーザーをユーザー登録画面へ飛ばす
-listing_user:自分の出品した商品の購入確認ページにはアクセスできないようにする

Commit:
カード情報編集用の部分テンプレートに変数でタイトルを表示

Commit:
カード情報変更機能の実装

Commit:
product_purchaseモデル作成

その他リンク等一部追加

ghost commented:
✅ Errors or warnings were resolved via the commit 770cf03! Please visit [Sider](https://sider.review/gh/repos/300881036/pulls/11) for details.

---

You can [turn off this notification](https://sider.review/gh/repos/300881036/settings#notifications) if you don't want to receive it from now on.

Commit:
商品購入機能の実装

購入ボタンを押すことでPAY.JPへ購入データを送信
アプリ内ではProduct_purchaseテーブルに購入情報を追加。
購入された商品については購入ボタンを押せない用に表示を切り替えるようにした。

Commit:
カード登録画面遷移時のturbolinksの無効化

turbolinksによりJSの読み込みが省略されフォームが表示されていなかった。
回避するためにturbolinksを無効化した。
その他細かい修正を行った。

Commit:
カード登録/更新画面のマークアップ

Commit:
fixup! カード登録/更新画面のマークアップ

Commit:
カード情報変更後の挙動調整

カード情報変更後、前のページへ戻るようにしました。

Commit:
モデルへのバリデーション

Commit:
カード登録画面の表示修正

エラーメッセージの表示を修正

Commit:
fixup! カード登録画面の表示修正

Commit:
fixup! fixup! カード登録画面の表示修正

Commit:
バリデーションの修正

Commit:
RSpecの導入

Commit:
カード未登録時は購入できないように設定

Commit:
factory_botの導入

Commit:
テストコードを記述

UserモデルとCardモデルについてテストコードを記述

Commit:
削除機能の実装（機能のみ）

マイページ未完成のため機能のみ実装。
リンクを貼ればすぐに使える状態。

nissy7ok committed:
Merge branch 'master' into develop_function_of_purchase

Commit:
購入完了画面の実装

CardForm、Confirmのクラスを流用しています。
見づらければ独立したSCSSファイルを作ります。

Commit:
コントローラーの修正

Commit:
Merge remote-tracking branch 'origin' into develop_function_of_purchase

Commit:
重複するマイグレーションファイルの統合/削除

Commit:
user.rb修正

nameという存在しないカラムが指定されていたため削除。
代わりにnicknameを作成したが入力画面が未実装のためコメントアウトしている。

Commit:
マイページ編集

カード情報の登録/変更機能実装
DBからユーザー情報を持ってくるように変更

Commit:
Siderの指摘事項を修正

Commit:
微修正

Commit:
fixup! 微修正

nissy7ok commented:
カード情報未登録では購入できない
<img width="952" alt="スクリーンショット 2020-11-13 22 24 01" src="https://user-images.githubusercontent.com/67008413/99077427-15a62e80-2600-11eb-8d07-258aabe64488.png">

カード情報登録画面
<img width="952" alt="スクリーンショット 2020-11-13 22 24 28" src="https://user-images.githubusercontent.com/67008413/99077438-1a6ae280-2600-11eb-9120-e23401304a58.png">

カード情報登録後は購入可能
<img width="952" alt="スクリーンショット 2020-11-13 22 24 36" src="https://user-images.githubusercontent.com/67008413/99077440-1b9c0f80-2600-11eb-9b91-2ed03bd7f10c.png">

マイページにはユーザー情報が表示されている
<img width="952" alt="スクリーンショット 2020-11-13 22 29 05" src="https://user-images.githubusercontent.com/67008413/99077445-1c34a600-2600-11eb-88ea-b25e6e27f05a.png">

カード情報登録済みの場合
<img width="952" alt="スクリーンショット 2020-11-13 22 29 11" src="https://user-images.githubusercontent.com/67008413/99077448-1ccd3c80-2600-11eb-8b89-e81447892c58.png">

カード情報未登録の場合
<img width="952" alt="スクリーンショット 2020-11-13 22 29 50" src="https://user-images.githubusercontent.com/67008413/99077450-1d65d300-2600-11eb-8ddb-1b009e6c06c1.png">

Commit:
削除機能実装

nissy7ok commented:
すみません削除機能忘れてました。実装しました。
<img width="952" alt="スクリーンショット 2020-11-13 22 49 50" src="https://user-images.githubusercontent.com/67008413/99078984-9f56fb80-2602-11eb-8b79-070235784b17.png">

Commit:
ビュー修正

fist_nameとlast_nameが逆でした。。。

yamato1491038 commented:
LGTM！

Commit:
Merge remote-tracking branch 'origin' into develop_function_of_purchase

Commit:
master更新のため再度コンフリクト解消

レイアウト変更に対応

yamagiti commented:
LGTM！！

yuta3817 commented:
LGTM

miyagit commented on app/controllers/items_controller.rb:
```diff
@@ -1,23 +1,88 @@
   end
 
   def mypage
+    @user = User.find_by(id: params[:id])
```
こちらfindメソッドを使いましょう。

nissy7ok replied:
ご指摘ありがとうございます。
```Hoge.find_by(id: params[:id])```
となっている部分をすべて
```Hoge.find(params[:id])```
に修正しました。

Commit:
find_by(id: params[:id]) => find(params[:id])へ修正

miyagit reviewed (approved):
レビュー依頼ありがとうございます、LGTMです！
